components/pkg-export-docker/src/cli.rs

```diff
@@ -0,0 +1,251 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(feature="clippy", feature(plugin))]
+#![cfg_attr(feature="clippy", plugin(clippy))]
+
+use clap::{App, Arg};
+use std::path::Path;
+use std::result;
+use std::str::FromStr;
+
+use hcore::package::PackageIdent;
+use url::Url;
+
+/// The version of this library and program when built.
+pub const VERSION: &'static str = include_str!(concat!(env!("OUT_DIR"), "/VERSION"));
+
+/// A Docker-specific clap:App wrapper
+#[derive(Clone)]
+pub struct Cli<'a, 'b>
+where
+    'a: 'b,
+{
+    pub app: App<'a, 'b>,
+}
+
+impl<'a, 'b> Cli<'a, 'b> {
+    pub fn new(name: &str, about: &'a str) -> Self {
+        Cli {
+            app: clap_app!(
+            (name) =>
+            (about: about)
+            (version: VERSION)
+            (author: "\nAuthors: The Habitat Maintainers <humans@habitat.sh>\n\n")
+            (@arg IMAGE_NAME: --("image-name") -i +takes_value
+                "Image name (default: \"{{pkg_origin}}/{{pkg_name}}\" supports: \
+                 {{pkg_origin}}, {{pkg_name}}, {{pkg_version}}, {{pkg_release}}, {{channel}})")
+            ),
+        }
+    }
+
+    pub fn add_base_packages_args(self) -> Self {
+        let app = self.app
+            .arg(
+                Arg::with_name("HAB_PKG")
+                    .long("hab-pkg")
+                    .value_name("HAB_PKG")
+                    .validator(valid_ident_or_hart)
+                    .help(
+                        "Habitat CLI package identifier (ex: acme/redis) or filepath to a Habitat \
+                        artifact (ex: /home/acme-redis-3.0.7-21120102031201-x86_64-linux.hart) to \
+                        install (default: core/hab)",
+                    ),
+            )
+            .arg(
+                Arg::with_name("HAB_LAUNCHER_PKG")
+                    .long("launcher-pkg")
+                    .value_name("HAB_LAUNCHER_PKG")
+                    .validator(valid_ident_or_hart)
+                    .help(
+                        "Launcher package identifier (ex: acme/redis) or filepath to a Habitat \
+                        artifact (ex: /home/acme-redis-3.0.7-21120102031201-x86_64-linux.hart) to \
+                        install (default: core/hab-launcher)",
+                    ),
+            )
+            .arg(
+                Arg::with_name("HAB_SUP_PKG")
+                    .long("sup-pkg")
+                    .value_name("HAB_SUP_PKG")
+                    .validator(valid_ident_or_hart)
```

Comment:
Same here but sure. :)

Reply:
Actually how do we achieve the first part (validating the identifier is for a valid supervisor/launcher)?

The second part should be easy since you just check the prefix of identifier string.

Reply:
this turned out to be not so trivial as I imagined since my idea was to do this in new custom validator for launcher and sup args but we don't have access to 'UI' in this part of the code yet. I created an issue for this and will look into it next week: https://github.com/habitat-sh/habitat/issues/4162 . Since other comments have been addressed already, I'd say this branch is read for merge from my side. :)